lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -322,6 +322,104 @@ void SPIRVRegularizeLLVMBase::expandSYCLTypeUsing(Module *M) {
     expandVIDWithSYCLTypeByValComp(F);
 }
 
+// In this function, we handle two conversion operations
+// 1. fptoui.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// 2. fptosi.sat.iX.fY (X is not 8,16,32,64; Y is 32 or 64)
+// Such non-standard integer types cannot be handled in SPIR-V. Hence, they
+// will be promoted to
+// 1. fptoui.sat.i64.fY (Y is 32 or 64)
+// 2. fptosi.sat.i64.fY (Y is 32 or 64)
+// However, LLVM documentation requires the following rules to be obeyed.
+// Rule 1: If the argument is any NaN, zero is returned.
+// Rule 2: If the argument is smaller than the smallest representable
+// (un)signed integer of the result type, the smallest representable
+// (un)signed integer is returned.
+// Rule 3: If the argument is larger than the largest representable (un)signed
+// integer of the result type, the largest representable (un)signed integer is
+// returned.
+// Rule 4: Otherwise, the result of rounding the argument towards zero is
+// returned.
+// Rules 1 & 4 are preserved when promoting iX to i64. For preserving Rule 2
+// and Rule 3, we saturate the result of the promoted instruction based on
+// original integer type (iX)
+// Example:
+// Input:
+// %0 = call i2 @llvm.fptosi.sat.i2.f32(float %input)
+// %1 = sext i32 %0
+// Output:
+// %0 = call i32 @_Z17convert_long_satf(float %input)
+// %1 = icmp sge i32 %0, 1 <Largest 2-bit signed integer>
+// %2 = icmp sle i32 %0, -2 <Smallest 2-bit signed integer>
+// %3 = select i1 %1, i32 1, i32 %0
+// %4 = select i1 %2, i32 -2, i32 %3
+// Replace uses of %1 in Input with %4 in Output
+void SPIRVRegularizeLLVMBase::cleanupConversionToNonStdIntegers(Module *M) {
+  for (auto I = M->begin(), E = M->end(); I != E;) {
+    Function *F = &(*I++);
+    std::vector<Instruction *> ToErase;
+    for (BasicBlock &BB : *F) {
+      for (Instruction &I : BB) {
```

Comment:
This can be done more efficiently.  Instead of iterating over all(!) instructions just to find some intrinsic calls, please consider iterating over all users of the intrinsics that need processing.

Reply:
There can be many intrinsics depending on the signature. But, I suppose we can surely try to parse over all function decls and see if any of them are fptosi.sat/fptoui.sat and process themaccordingly.
Can I please add a TODO for this optimization and address that in a later PR? Thanks for this suggestion.

Reply:
> But, I suppose we can surely try to parse over all function decls

There is no need for parsing. You already have the `Function` object, so you can call `F->getIntrinsicID()`.  If the IntrinsicID matches what you're interested in, you then iterate over `F->users()`.

> Can I please add a TODO for this optimization and address that in a later PR? Thanks for this suggestion.

If you don't intend to rework the loops in this PR, please at least add an early exit based on `F->getIntrinsicID()` in the outer loop to avoid every translation needlessly iterating over all instructions.

Reply:
I think early exit may not be applicable here as we want to parse functions that contain uses of such intrinsics.

I might have over-thought the resolution here though. You make a very good point. We do already have a handle to F that we can use here. I will try to put up a change here. 

Thanks so much again for feedback.

Reply:
Awesome, thanks for addressing this!